provider/keycloak_openid_user_attribute_protocol_mapper.go

```diff
@@ -0,0 +1,229 @@
+	return &keycloak.OpenIdUserAttributeProtocolMapper{
+		Id:               data.Id(),
+		Name:             data.Get("name").(string),
+		RealmId:          data.Get("realm_id").(string),
+		ClientId:         data.Get("client_id").(string),
+		ClientScopeId:    data.Get("client_scope_id").(string),
+		AddToIdToken:     data.Get("add_to_id_token").(bool),
+		AddToAccessToken: data.Get("add_to_access_token").(bool),
+		AddToUserInfo:    data.Get("add_to_user_info").(bool),
+
+		UserAttribute:  data.Get("user_attribute").(string),
+		ClaimName:      data.Get("claim_name").(string),
+		ClaimValueType: data.Get("claim_value_type").(string),
+		Multivalued:    data.Get("multivalued").(bool),
+	}
+}
+
+func mapFromOpenIdUserAttributeMapperToData(mapper *keycloak.OpenIdUserAttributeProtocolMapper, data *schema.ResourceData) {
+	data.SetId(mapper.Id)
+	data.Set("name", mapper.Name)
+	data.Set("realm_id", mapper.RealmId)
+
+	if mapper.ClientId != "" {
+		data.Set("client_id", mapper.ClientId)
+	} else {
+		data.Set("client_scope_id", mapper.ClientScopeId)
+	}
+
+	data.Set("add_to_id_token", mapper.AddToIdToken)
+	data.Set("add_to_access_token", mapper.AddToAccessToken)
+	data.Set("add_to_user_info", mapper.AddToUserInfo)
+	data.Set("user_attribute", mapper.UserAttribute)
+	data.Set("claim_name", mapper.ClaimName)
+	data.Set("claim_value_type", mapper.ClaimValueType)
+	data.Set("multivalued", mapper.Multivalued)
+}
+
+func resourceKeycloakOpenIdUserAttributeProtocolMapperCreate(data *schema.ResourceData, meta interface{}) error {
+	clientId := data.Get("client_id")
+	clientScopeId := data.Get("client_scope_id")
+
+	if clientId == "" && clientScopeId == "" {
+		return fmt.Errorf("one of client_id or client_scope_id must be set")
+	}
+
+	keycloakClient := meta.(*keycloak.KeycloakClient)
+
+	openIdUserAttributeMapper := mapFromDataToOpenIdUserAttributeProtocolMapper(data)
+
+	err := keycloakClient.NewOpenIdUserAttributeProtocolMapper(openIdUserAttributeMapper)
+
+	if err != nil {
+		return err
+	}
+
+	mapFromOpenIdUserAttributeMapperToData(openIdUserAttributeMapper, data)
+
+	return resourceKeycloakOpenIdUserAttributeProtocolMapperRead(data, meta)
+}
+
+func resourceKeycloakOpenIdUserAttributeProtocolMapperRead(data *schema.ResourceData, meta interface{}) error {
+	keycloakClient := meta.(*keycloak.KeycloakClient)
+
+	realmId := data.Get("realm_id").(string)
+	clientId := data.Get("client_id").(string)
+	clientScopeId := data.Get("client_scope_id").(string)
+
+	var openIdUserAttributeMapper *keycloak.OpenIdUserAttributeProtocolMapper
+	var err error
+	if clientId != "" {
+		openIdUserAttributeMapper, err = keycloakClient.GetOpenIdUserAttributeProtocolMapperForClient(realmId, clientId, data.Id())
+	} else {
+		openIdUserAttributeMapper, err = keycloakClient.GetOpenIdUserAttributeProtocolMapperForClientScope(realmId, clientScopeId, data.Id())
+	}
```

Comment:
I wonder if this would look better if there was just a single function, `keycloakClient.GetOpenIdUserAttributeProtocolMapper`, that took both client id and client scope id, and figured out which of the two this mapper belongs to based on which string is empty.  I'd rather have the code in the `keycloak` package do stuff like this.